internal/sync/syncer_related_test.go

```diff
@@ -0,0 +1,367 @@
+/*
+Copyright 2025 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+	"go.uber.org/zap"
+
+	dummyv1alpha1 "github.com/kcp-dev/api-syncagent/internal/sync/apis/dummy/v1alpha1"
+	"github.com/kcp-dev/api-syncagent/internal/test/diff"
+	syncagentv1alpha1 "github.com/kcp-dev/api-syncagent/sdk/apis/syncagent/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/kontext"
+)
+
+func newPublishedResources(relatedResources []syncagentv1alpha1.RelatedResourceSpec) *syncagentv1alpha1.PublishedResource {
+	return &syncagentv1alpha1.PublishedResource{
+		Spec: syncagentv1alpha1.PublishedResourceSpec{
+			Resource: syncagentv1alpha1.SourceResourceDescriptor{
+				APIGroup: dummyv1alpha1.GroupName,
+				Version:  dummyv1alpha1.GroupVersion,
+				Kind:     "NamespacedThing",
+			},
+			Projection: &syncagentv1alpha1.ResourceProjection{
+				Kind: "RemoteThing",
+			},
+			Naming: &syncagentv1alpha1.ResourceNaming{
+				Name: "$remoteClusterName-$remoteName",
+			},
+			Related: relatedResources,
+		},
+	}
+}
+
+func TestSyncerProcessingRelatedResources(t *testing.T) {
+	const stateNamespace = "kcp-system"
+
+	type testcase struct {
+		name                 string
+		remoteAPIGroup       string
+		localCRD             *apiextensionsv1.CustomResourceDefinition
+		pubRes               *syncagentv1alpha1.PublishedResource
+		remoteObject         *unstructured.Unstructured
+		localObject          *unstructured.Unstructured
+		existingState        string
+		performRequeues      bool
+		expectedRemoteObject *unstructured.Unstructured
+		expectedLocalObject  *unstructured.Unstructured
```

Comment:
If we're testing related resources, these should be *corev1.Secret and ensure the relatd object, not the main object. No need to re-test what other unit tests have already tested.